x/mongo/driver/topology/connection.go

```diff
@@ -48,6 +48,8 @@ type connection struct {
 	// - atomic bug: https://pkg.go.dev/sync/atomic#pkg-note-BUG
 	// - suggested layout: https://go101.org/article/memory-layout.html
 	state int64
+	inUse bool
+	err   error
```

Comment:
I don't see a concurrent issue in the current code. The `inUse` is set in checkin and checkout, and the value is only read by interruption between checkin and checkout.

The `err` is optional merely to indicate the `ConnectionError` from reading/writing more clearly.

Reply:
The `clear` or `clearAll` method may be called concurrently with `checkIn`/`checkOut`, leading to a concurrent read/write of `inUse`.

Some possible solutions:
1. Make `inUse` an `atomic.Bool`.
2. Keep a set of in-use connections, similar to how we keep a set of all connections and idle connections.
3. Close all connections if `interruptInUseConnections=true` (i.e. don't try to distinguish between in-use and idle).

(1) is the smallest change from the current implementation. However, I think we should seriously consider (3) because it doesn't seem like the runtime optimization of lazy-closing idle connections (when we're about to force-close in-use connections) is worth the significant increase in code complexity necessary to keep track of what connections are in-use.